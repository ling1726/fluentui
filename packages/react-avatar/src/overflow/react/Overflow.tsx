import * as React from 'react';
import { useMergedRefs } from '@fluentui/react-utilities';
import { OverflowGroupState, OnUpdateOverflow, OverflowEventPayload } from '../native/overflowManager';
import { OverflowContext } from './overflowContext';
import { useOverflowContainer, UseOverflowContainerOptions } from './useOverflowContainer';

export type OverflowChildProps = React.HTMLAttributes<HTMLElement> & {
  ref?: React.Ref<never>;
};

export type OverflowProps = UseOverflowContainerOptions & {
  children: React.ReactElement;
};

export const Overflow = React.forwardRef((props: OverflowProps, ref) => {
  const [hasOverflow, setHasOverflow] = React.useState(false);
  const [itemVisibility, setItemVisibility] = React.useState<Record<string, boolean>>({});
  const [groupVisibility, setGroupVisibility] = React.useState<Record<string, OverflowGroupState>>({});

  const updateItemVisibility: OnUpdateOverflow = (data: OverflowEventPayload) => {
    setHasOverflow(() => data.invisibleItems.length > 0);
    setItemVisibility(() => {
      const newState: Record<string, boolean> = {};
      data.visibleItems.forEach(x => (newState[x.id] = true));
      data.invisibleItems.forEach(x => (newState[x.id] = false));
      return newState;
    });
    setGroupVisibility(groupVisibility);
  };

  const { containerRef, registerItem, updateOverflow } = useOverflowContainer(updateItemVisibility, {
    ...props,
  });

  const mergedRef = useMergedRefs(containerRef, ref);

  const children = React.cloneElement(React.Children.only(props.children), { ref: mergedRef });

  return (
    <OverflowContext.Provider
      value={{
        itemVisibility,
        groupVisibility,
        hasOverflow,
        registerItem,
        updateOverflow,
      }}
    >
      {children}
    </OverflowContext.Provider>
  );
});
