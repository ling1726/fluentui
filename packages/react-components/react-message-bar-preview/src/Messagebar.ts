export * from './components/Messagebar/index';
