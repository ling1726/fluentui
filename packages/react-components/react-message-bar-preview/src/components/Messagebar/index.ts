export * from './Messagebar';
export * from './Messagebar.types';
export * from './renderMessagebar';
export * from './useMessagebar';
export * from './useMessagebarStyles.styles';
