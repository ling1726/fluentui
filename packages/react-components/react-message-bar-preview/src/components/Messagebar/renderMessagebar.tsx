/** @jsxRuntime automatic */
/** @jsxImportSource @fluentui/react-jsx-runtime */

import { assertSlots } from '@fluentui/react-utilities';
import type { MessagebarState, MessagebarSlots } from './Messagebar.types';

/**
 * Render the final JSX of Messagebar
 */
export const renderMessagebar_unstable = (state: MessagebarState) => {
  assertSlots<MessagebarSlots>(state);

  // TODO Add additional slots in the appropriate place
  return <state.root />;
};
