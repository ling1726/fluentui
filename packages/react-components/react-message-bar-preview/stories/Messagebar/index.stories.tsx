import { Messagebar } from '@fluentui/react-message-bar-preview';

import descriptionMd from './MessagebarDescription.md';
import bestPracticesMd from './MessagebarBestPractices.md';

export { Default } from './MessagebarDefault.stories';

export default {
  title: 'Preview Components/Messagebar',
  component: Messagebar,
  parameters: {
    docs: {
      description: {
        component: [descriptionMd, bestPracticesMd].join('\n'),
      },
    },
  },
};
