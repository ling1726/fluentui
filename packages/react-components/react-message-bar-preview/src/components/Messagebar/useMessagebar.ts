import * as React from 'react';
import { getNativeElementProps, slot } from '@fluentui/react-utilities';
import type { MessagebarProps, MessagebarState } from './Messagebar.types';

/**
 * Create the state required to render Messagebar.
 *
 * The returned state can be modified with hooks such as useMessagebarStyles_unstable,
 * before being passed to renderMessagebar_unstable.
 *
 * @param props - props from this instance of Messagebar
 * @param ref - reference to root HTMLElement of Messagebar
 */
export const useMessagebar_unstable = (props: MessagebarProps, ref: React.Ref<HTMLElement>): MessagebarState => {
  return {
    // TODO add appropriate props/defaults
    components: {
      // TODO add each slot's element type or component
      root: 'div',
    },
    // TODO add appropriate slots, for example:
    // mySlot: resolveShorthand(props.mySlot),
    root: slot.always(
      getNativeElementProps('div', {
        ref,
        ...props,
      }),
      { elementType: 'div' },
    ),
  };
};
