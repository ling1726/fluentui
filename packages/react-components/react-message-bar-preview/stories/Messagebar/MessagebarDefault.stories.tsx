import * as React from 'react';
import { Messagebar, MessagebarProps } from '@fluentui/react-message-bar-preview';

export const Default = (props: Partial<MessagebarProps>) => <Messagebar {...props} />;
