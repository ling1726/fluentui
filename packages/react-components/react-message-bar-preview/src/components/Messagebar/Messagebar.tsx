import * as React from 'react';
import type { ForwardRefComponent } from '@fluentui/react-utilities';
import { useMessagebar_unstable } from './useMessagebar';
import { renderMessagebar_unstable } from './renderMessagebar';
import { useMessagebarStyles_unstable } from './useMessagebarStyles.styles';
import type { MessagebarProps } from './Messagebar.types';

/**
 * Messagebar component - TODO: add more docs
 */
export const Messagebar: ForwardRefComponent<MessagebarProps> = React.forwardRef((props, ref) => {
  const state = useMessagebar_unstable(props, ref);

  useMessagebarStyles_unstable(state);
  return renderMessagebar_unstable(state);
});

Messagebar.displayName = 'Messagebar';
