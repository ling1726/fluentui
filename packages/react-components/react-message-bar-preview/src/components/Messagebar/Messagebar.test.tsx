import * as React from 'react';
import { render } from '@testing-library/react';
import { isConformant } from '../../testing/isConformant';
import { Messagebar } from './Messagebar';

describe('Messagebar', () => {
  isConformant({
    Component: Messagebar,
    displayName: 'Messagebar',
  });

  // TODO add more tests here, and create visual regression tests in /apps/vr-tests

  it('renders a default state', () => {
    const result = render(<Messagebar>Default Messagebar</Messagebar>);
    expect(result.container).toMatchSnapshot();
  });
});
