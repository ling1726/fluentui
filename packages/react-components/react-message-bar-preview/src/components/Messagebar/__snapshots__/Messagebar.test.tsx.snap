// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Messagebar renders a default state 1`] = `
<div>
  <div
    class="fui-Messagebar"
  >
    Default Messagebar
  </div>
</div>
`;
